docs/tech/aca.md

```diff
@@ -0,0 +1,119 @@
+# Attribute Certificate Authority
+
+To support Attribute Based Access Control (ABAC) the user has to pass a set of attributes during TCert creation, these attributes will be used during transaction execution or query to determine whether the user can or cannot execute a specific chaincode. A mechanism to validate the ownership of attributes is required in order to prove if the attributes passed by the user are correct. The Attribute Certificate Authority (ACA) has the responsibility of validating attributes and in response it returns an Attribute Certificate (ACert) with the valid attribute values.
+
+ACA will have a database to hold attributes for each user and affiliation. This DB will be encrypted.
+
+1. userID _(The id passed by the user during enrollment)_
+2. affiliation _(The entity which the user is affiliated to)_
+3. attribute Key _(The key used to look for the attribute, e.g. 'position')_
+4. attribute value _(The value of the attribute, e.g. 'software engineer')_
+5. valid from _(The start of the attribute's validity period)_
+6. valid to _(The end of the attribute's validity period)_
+
+## FLOW
+
+![ACA flow](https://raw.githubusercontent.com/andresgaragiola/fabric/ABAC/docs/images/aca_flow.jpg)
+
+### During enrollment
+
+1. The user requests an Enrollment Certificate (ECert) to ECA
+2. ECA creates the ECert and responds to the user with it.
+3. ECA issues a fetch request under TLS to the ACA passing the newly generated ECert as a parameter. This request is signed with the ECA's private key.
+4. The request triggers ACA asynchronous mechanism that fetches attributes' values from external sources and populates the attributes database (in the current implementation attributes are loaded from an internal configuration file).
+
+### During TCert generation
+
+1. When the user needs TCerts to create a new transaction it requests a batch of TCerts to the TCA, and provides the following:
+   * The batch size (i.e. how many TCerts the user is expecting)
+   * Its ECert
+   * A list of attributes (e.g. Company:IBM, Position: Software Engineer)
+2. Under TLS TCA sends a RequestAttributes() to ACA to verify if the user is in possession of those attributes. This request is signed with TCA's private key and it contains:
+   * User's ECert
+   * A key/pair list of attributes where the value of the attribute is hashed:
+     * Company: Hash(IBM)
+     * Position: Hash(Software Engineer)
```

Comment:
The main reason is force ACA to return its internal value. If the value is sent we didn't know if the ACA response corresponds to the passed value or the internal value. In this way the TCA pass a hash of the value an ACA is forced to calculate the same hash from its internal value to compare.